domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+func NewService(st State) *Service {
+	return &Service{st: st}
+}
+
+// GetUser will find and return the user associated with name. If there is no
+// user for the user name then an error that satisfies usererrors.NotFound will
+// be returned. If supplied with a invalid user name then a error that satisfies
+// usererrors.
+//
+// GetUser will not return users that have been previously deleted or removed.
+func (s *Service) GetUser(
+	ctx context.Context,
+	name string,
+) (user.User, error) {
+	if err := ValidateUsername(name); err != nil {
+		return user.User{}, fmt.Errorf("username %q: %w", name, err)
+	}
+
+	u, err := s.st.GetUser(ctx, name)
+	if err != nil {
+		return user.User{}, fmt.Errorf("getting user %q: %w", name, err)
+	}
+
+	return u, nil
+}
+
+// ValidateUsername takes a user name and validates that the user name is
+// conformant to our regex rules defined in usernameValidationRegex. If a user
+// name is not valid a error is returned that satisfies
+// usererrors.UsernameNotValid.
+//
+// Usernames must be 1 or more runes long, can contain any unicode rune from the
+// letter class and may contain zero or more of .,+ or - runes as long as they
+// are don't appear at the start or end of the username.
+func ValidateUsername(name string) error {
+	regex, err := regexp.Compile(usernameValidationRegex)
+	if err != nil {
+		return fmt.Errorf("compiling user name validation regex %q: %w",
+			usernameValidationRegex, err,
+		)
+	}
+	if !regex.MatchString(name) {
+		return fmt.Errorf("%w %q", usererrors.UsernameNotValid, name)
+	}
+	return nil
+}
+
+// AddUser will add a new user to the database. If the user already exists an
+// error that satisfies usererrors.AlreadyExists will be returned.
+func (s *Service) AddUser(ctx context.Context, name, displayName, password, creator string) error {
+	err := ValidateUsername(name)
+	if err != nil {
+		return fmt.Errorf("username %q: %w", name, err)
+	}
+
+	err = s.st.AddUser(ctx, name, displayName, password, creator)
+	if err != nil {
+		return fmt.Errorf("adding user %q: %w", name, err)
+	}
+	return nil
+}
+
+// RemoveUser marks the user as deleted. This obviates the ability of a user
+// to function, but keeps the userDoc retaining provenance, i.e. auditing.
+func (s *Service) RemoveUser(ctx context.Context, name string) error {
+	if err := ValidateUsername(name); err != nil {
+		return fmt.Errorf("username %q: %w", name, err)
+	}
+	err := s.st.RemoveUser(ctx, name)
+	if err != nil {
+		return fmt.Errorf("removing user %q: %w", name, err)
+	}
+	return nil
+}
+
+// SetPassword removes any active activation keys and sets the user password
+func (s *Service) SetPassword(ctx context.Context, name string,
+	password user.UserPassword) error {
+	if err := validateUserName(name); err != nil {
+		return fmt.Errorf("username %q: %w", name, err)
+	}
+	lowercaseName := strings.ToLower(name)
+	if err := s.st.RemoveUserActivationKey(ctx, lowercaseName); err != nil {
```

Comment:
Set password in the state layer should remove any activation keys that way it happens all in the one transaction.

We don't want to have it as two operations in the services layer otherwise we risk ending up in a bad state.